src/Aspire.Hosting.Azure.ServiceBus/AzureServiceBusExtensions.cs

```diff
@@ -122,21 +132,373 @@ public static IResourceBuilder<AzureServiceBusResource> AddQueue(this IResourceB
     /// <param name="topicName">The name of the topic.</param>
     /// <param name="subscriptionName">The name of the subscription.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    [Obsolete($"This method is obsolete and will be removed in a future version. Use {nameof(WithTopic)} instead to add an Azure Service Bus Subscription to a Topic.")]
     public static IResourceBuilder<AzureServiceBusResource> AddSubscription(this IResourceBuilder<AzureServiceBusResource> builder, string topicName, string subscriptionName)
     {
-        builder.Resource.Subscriptions.Add((topicName, subscriptionName));
+        builder.WithTopic(topicName, topic =>
+        {
+            if (!topic.Subscriptions.Any(x => x.Name == subscriptionName))
+            {
+                topic.Subscriptions.Add(new ServiceBusSubscription(subscriptionName));
+            }
+        });
+
+        return builder;
+    }
+
+    /// <summary>
+    /// Configures an Azure Service Bus resource to be emulated. This resource requires an <see cref="AzureServiceBusResource"/> to be added to the application model.
+    /// </summary>
+    /// <remarks>
+    /// This version of the package defaults to the <inheritdoc cref="ServiceBusEmulatorContainerImageTags.Tag"/> tag of the <inheritdoc cref="ServiceBusEmulatorContainerImageTags.Registry"/>/<inheritdoc cref="ServiceBusEmulatorContainerImageTags.Image"/> container image.
+    /// </remarks>
+    /// <param name="builder">The Azure Service Bus resource builder.</param>
+    /// <param name="configureContainer">Callback that exposes underlying container used for emulation to allow for customization.</param>
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    /// <example>
+    /// The following example creates an Azure Service Bus resource that runs locally is an emulator and referencing that
+    /// resource in a .NET project.
+    /// <code lang="csharp">
+    /// var builder = DistributedApplication.CreateBuilder(args);
+    ///
+    /// var serviceBus = builder.AddAzureServiceBus("myservicebus")
+    ///    .RunAsEmulator()
+    ///    .AddQueue("queue");
+    ///
+    /// builder.AddProject&lt;Projects.InventoryService&gt;()
+    ///        .WithReference(serviceBus);
+    ///
+    /// builder.Build().Run();
+    /// </code>
+    /// </example>
+    public static IResourceBuilder<AzureServiceBusResource> RunAsEmulator(this IResourceBuilder<AzureServiceBusResource> builder, Action<IResourceBuilder<AzureServiceBusEmulatorResource>>? configureContainer = null)
+    {
+        if (builder.ApplicationBuilder.ExecutionContext.IsPublishMode)
+        {
+            return builder;
+        }
+
+        // Add emulator container
+        var configHostFile = Path.Combine(Path.GetTempPath(), "AspireServiceBusEmulator", Path.GetRandomFileName() + ".json");
+
+        Directory.CreateDirectory(Path.GetDirectoryName(configHostFile)!);
```

Comment:
We should be using `Directory.CreateTempSubdirectory` here - https://learn.microsoft.com/en-us/dotnet/api/system.io.directory.createtempsubdirectory?view=net-9.0. The temp directory isn't guaranteed to be private to the current user.

Reply:
Yep, we use it all over aspire.

Reply:
Done